### PR TITLE
Cleanups before Release

### PR DESCRIPTION
**Pull request by JesusMcCloud**
Proper version of #82 going into development.  
Once merged, will prepare a release.

Changes the following:

 * Rename `toCryptoAlgorithm` to `toX509SignatureAlgorithm` accordingly
 * Rework CryptoSignature to two-dimensional interface:
   * CryptoSignature <- {EC <- {IndefiniteLength, DefiniteLength}, RsaOrHmac}
    * CryptoSignature <- {Defined <- {Ill <- EC.IndefiniteLength, Well <- {EC.DefiniteLength, RsaOrHmac}}

This finally makes it possible to get a common ground to encode RSA/HMAC and EC signatures for COSE and JWS, without having to jump through hoops. The effect of this can be seen in [the corresponding  VC Lib PR](https://github.com/a-sit-plus/kmm-vc-library/pull/83).

### CODE REVIEW

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt [hunk 1]
```diff
@@ -6,54 +6,59 @@ import at.asitplus.crypto.datatypes.asn1.DERTags.DER_SEQUENCE
...
-    abstract val rawByteArray: ByteArray
-
-    abstract fun encodeToTlvBitString(): Asn1Element
-
-    override fun equals(other: Any?): Boolean {
-        if (this === other) return true
-        if (other == null || this::class != other::class) return false
-
-        other as CryptoSignature
+    sealed interface Defined : CryptoSignature {
```
Comment (iaik-jheher) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt:36]: what's the benefit of this inheritance structure? why not just `IllDefined: CryptoSignature` and `WellDefined: CryptoSignature`? (Or even drop `IllDefined` altogether since it seems to add nothing?)

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/ECPoint.kt [hunk 2]
```diff
@@ -87,7 +90,11 @@ sealed class ECPoint private constructor(
         else if (this is Normalized)
             "ECPoint[$curve]: (${homX.toString(16)} : ${homY.toString(16)}) [normalized]"
         else
-            "ECPoint[$curve]: (${(homX/homZ).toString(16)} : ${(homY/homZ).toString(16)}) [with Z = ${homZ.toString(16)}]"
+            "ECPoint[$curve]: (${(homX / homZ).toString(16)} : ${(homY / homZ).toString(16)}) [with Z = ${
+                homZ.toString(
+                    16
+                )
+            }]"
```
Comment (iaik-jheher) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/ECPoint.kt:97]: this indentation change feels particularly excessive (and quite a few of the others make the code less readable imo, too)

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/misc/PointCompression.kt [hunk 3]
```diff
@@ -11,29 +11,43 @@ enum class ANSIECPrefix(val prefixByte: Byte) {
...
     val prefixUByte inline get() = prefixByte.toUByte()
 
+    @Suppress("NOTHING_TO_INLINE")
     inline operator fun plus(that: ByteArray) = byteArrayOf(prefixByte, *that)
 
     companion object {
+
+        /**
+         * Gets the ANSI prefix for [byte].
+         * @throws IllegalArgumentException for ZERO
```
Comment (iaik-jheher) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/misc/PointCompression.kt:29]: copy/paste error?

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/misc/PointCompression.kt [hunk 4]
```diff
@@ -68,9 +82,9 @@ internal fun decompressY(curve: ECCurve, x: ModularBigInteger, sign: Sign): Modu
     val alpha = x.pow(3) + curve.a * x + curve.b
 
     require(quadraticResidueTest(alpha))
-        { "Invalid compressed point (x=$x) on $curve"}
+    { "Invalid compressed point (x=$x) on $curve" }
```
Comment (iaik-jheher) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/misc/PointCompression.kt:85]: i really dislike removing the indentation here

File: CHANGELOG.md [hunk 5]
```diff
@@ -139,17 +139,26 @@
...
 
 **Fixes**
  * Move `curve` from `CryptoAlgorithm` to `JwsAlgorithm`
- * Don't read curve information into the X.509 tea leaves where none exists
+ * Don't assume curve information for the X.509 signature when, in fact, none exists
    * `CryptoSignature`s in X.509 are now indefinite length
- * Rename `CryptoAlgorithm` to `X509SignatureAlgorithm` to better describe what it is
+
+**Changes**
+* Drop support for did-encoding uncompressed keys (but keep decoding support)
```
Comment (iaik-jheher) [inline CHANGELOG.md:159]:
```suggestion
* Always DID-encode keys in compressed form (but keep decoding support)
```

File: datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKeyParams.kt [hunk 6]
```diff
@@ -89,13 +87,13 @@ sealed class CoseKeyParams {
         override fun yHashCode(): Int = y?.contentHashCode() ?: 0
 
         override fun toCryptoPublicKey(): KmmResult<CryptoPublicKey> {
-            return runCatching {
-                CryptoPublicKey.EC(
+            return catching {
```
Comment (nodh) [inline datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKeyParams.kt:90]: Probably makes sense to convert from block body to expression

File: datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKeyParams.kt [hunk 7]
```diff
@@ -167,13 +164,13 @@ sealed class CoseKeyParams {
         }
 
         override fun toCryptoPublicKey(): KmmResult<CryptoPublicKey> {
-            return runCatching {
+            return catching {
```
Comment (nodh) [inline datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKeyParams.kt:167]: Probably makes sense to convert from block body to expression

File: datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseSigned.kt [hunk 8]
```diff
@@ -87,14 +82,14 @@ data class CoseSigned(
...
+        fun deserialize(it: ByteArray) = catching {
             cborSerializer.decodeFromByteArray<CoseSigned>(it)
-        }.wrap()
+        }
     }
 }
 
 fun CoseHeader.usesEC(): Boolean? {
-    algorithm?.toCryptoAlgorithm()?.isEc?.let { return it }
+    algorithm?.toX509SignatureAlgorithm()?.isEc?.let { return it }
```
Comment (nodh) [inline datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseSigned.kt:92]: Probably can be refactored to an expression body

File: datatypes-jws/src/commonMain/kotlin/at/asitplus/crypto/datatypes/jws/JsonWebKey.kt [hunk 9]
```diff
@@ -281,10 +282,10 @@ data class JsonWebKey(
      * (i.e. if all key params are set), or the first error.
      */
     fun toCryptoPublicKey(): KmmResult<CryptoPublicKey> =
-        runCatching {
+        catching {
```
Comment (nodh) [inline datatypes-jws/src/commonMain/kotlin/at/asitplus/crypto/datatypes/jws/JsonWebKey.kt:285]: This `catching` can probably moved up one line, to reduce indendation in the body

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoPublicKey.kt [hunk 10]
```diff
@@ -347,24 +351,34 @@ sealed class CryptoPublicKey : Asn1Encodable<Asn1Sequence>, Identifiable {
...
+            )
+            @Suppress("NOTHING_TO_INLINE")
             inline operator fun invoke(curve: ECCurve, x: ByteArray, usePositiveY: Boolean) =
                 fromCompressed(curve, x, usePositiveY)
 
-            @Deprecated("Explicitly specify what you want", ReplaceWith(
-                "fromUncompressed(curve, x, y)"))
+            @Deprecated(
+                "Explicitly specify what you want", ReplaceWith(
+                    "fromUncompressed(curve, x, y)"
```
Comment (nodh) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoPublicKey.kt:378]: Why is this wrapped here?

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt [hunk 11]
```diff
@@ -6,54 +6,59 @@ import at.asitplus.crypto.datatypes.asn1.DERTags.DER_SEQUENCE
...
+        }
     }
 
-    override fun hashCode(): Int = signature.hashCode()
+    fun encodeToTlvBitString(): Asn1Element
 
     override fun encodeToTlv(): Asn1Element = signature
 
-    override fun toString(): String {
+    fun str(): String {
```
Comment (nodh) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt:61]: Method name seems a bit short, and can probably be converted into an expression

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/X509SignatureAlgorithm.kt [hunk 12]
```diff
@@ -11,18 +12,9 @@ import kotlinx.serialization.descriptors.SerialDescriptor
...
- *
- * This constant lives here, because we also need it in the commons module to be able to map this JWS Algorithm to a CryptoAlgorithm.
- * (It cannot be put into the compilation, since it is needed for enum init).
- */
-val OID_ECDH_ES = ObjectIdentifier("1.3.132.1.12")
-
 @Serializable(with = X509SignatureAlgorithmSerializer::class)
-enum class X509SignatureAlgorithm(override val oid: ObjectIdentifier, val isEc: Boolean = false)
-    : Asn1Encodable<Asn1Sequence>, Identifiable {
+enum class X509SignatureAlgorithm(override val oid: ObjectIdentifier, val isEc: Boolean = false) :
```
Comment (nodh) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/X509SignatureAlgorithm.kt:16]: Maybe wrap this the other way round?

**Review by iaik-jheher — CHANGES_REQUESTED** [review]
.

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/io/Encoding.kt [hunk 13]
```diff
@@ -50,7 +51,7 @@ object ByteArrayBase64Serializer : KSerializer<ByteArray> {
      * @throws SerializationException on error
      */
     override fun deserialize(decoder: Decoder): ByteArray {
-        return kotlin.runCatching { decoder.decodeString().decodeToByteArray(Base64Strict) }
+        return catching { decoder.decodeString().decodeToByteArray(Base64Strict) }
```
Comment (nodh) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/io/Encoding.kt:54]: can be converted to an expression

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/io/Encoding.kt [hunk 14]
```diff
@@ -74,7 +75,7 @@ object ByteArrayBase64UrlSerializer : KSerializer<ByteArray> {
      * @throws SerializationException on error
      */
     override fun deserialize(decoder: Decoder): ByteArray {
-        return kotlin.runCatching { decoder.decodeString().decodeToByteArray(Base64UrlStrict) }
+        return catching { decoder.decodeString().decodeToByteArray(Base64UrlStrict) }
```
Comment (nodh) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/io/Encoding.kt:78]: can be converted to an expression

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/misc/PointCompression.kt [hunk 15]
```diff
@@ -68,9 +82,9 @@ internal fun decompressY(curve: ECCurve, x: ModularBigInteger, sign: Sign): Modu
...
     require(curve.modulus.bitAt(0) && curve.modulus.bitAt(1)) // (modulus % 4) == 3
-        { "Decompression on $curve requires Tonelli-Shanks Algorithm" }
+    { "Decompression on $curve requires Tonelli-Shanks Algorithm" }
```
Comment (nodh) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/misc/PointCompression.kt:87]: Same with the indendation here

File: datatypes/src/jvmMain/kotlin/at/asitplus/crypto/datatypes/JcaExtensions.kt [hunk 16]
```diff
@@ -80,43 +80,43 @@ fun CryptoPublicKey.getJcaPublicKey() = when (this) {
 }
 
 fun CryptoPublicKey.EC.getJcaPublicKey(): KmmResult<ECPublicKey> {
-    return runCatching {
+    return catching {
```
Comment (nodh) [inline datatypes/src/jvmMain/kotlin/at/asitplus/crypto/datatypes/JcaExtensions.kt:83]: can be converted to an expression

File: datatypes/src/jvmMain/kotlin/at/asitplus/crypto/datatypes/JcaExtensions.kt [hunk 17]
```diff
@@ -80,43 +80,43 @@ fun CryptoPublicKey.getJcaPublicKey() = when (this) {
...
         JCEECPublicKey("EC", ecPublicKeySpec)
-    }.wrap()
+    }
 }
 
 private val rsaFactory = KeyFactory.getInstance("RSA")
 
 fun CryptoPublicKey.Rsa.getJcaPublicKey(): KmmResult<RSAPublicKey> =
-    runCatching {
+    catching {
```
Comment (nodh) [inline datatypes/src/jvmMain/kotlin/at/asitplus/crypto/datatypes/JcaExtensions.kt:96]: move up this `catching` a line to reduce indendation in the body?

File: datatypes/src/jvmTest/kotlin/at/asitplus/crypto/datatypes/X509CertificateJvmTest.kt [hunk 18]
```diff
@@ -82,7 +83,7 @@ class X509CertificateJvmTest : FreeSpec({
             initSign(keyPair.private)
             update(tbsCertificate.encodeToTlv().derEncoded)
         }.sign()
-        val test = (CryptoSignature.decodeFromDer(signed) as CryptoSignature.EC.IndefiniteLength).withCurve(ECCurve.SECP_256_R_1)
+        val test =            CryptoSignature.decodeFromDer(signed)
```
Comment (nodh) [inline datatypes/src/jvmTest/kotlin/at/asitplus/crypto/datatypes/X509CertificateJvmTest.kt:86]: spaces here seem off

**Review by nodh — CHANGES_REQUESTED** [review]

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt (hunk 1 above)
Comment (JesusMcCloud) [reply datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt:36]: Without two mutually exlcusive Types, you won't win anything, so it needs to stay, because otherwise

this
```
//it is CryptoSignature.Defined
when (it) {
    is CryptoSignature.Defined.Well -> it
    is CryptoSignature.Defined.Ill -> when (it) {
        is CryptoSignature.EC.IndefiniteLength -> it.withCurve((publicKey as CryptoPublicKey.EC).curve)
    }
}
```

turns into this useless mess:
```
//it is CryptoSignature
when (it) {
    is CryptoSignature.Definite -> it
    else -> when (it) {
        is CryptoSignature.EC.IndefiniteLength -> it.withCurve((publicKey as CryptoPublicKey.EC).curve)
        is CryptoSignature.EC.DefiniteLength -> TODO("utterly useless")
        is CryptoSignature.RSAorHMAC -> TODO("utterly useless")
    }
}
```

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoPublicKey.kt (hunk 10 above)
Comment (JesusMcCloud) [reply datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoPublicKey.kt:378]: let me rephrase that for you: what's the deal with the auto-formatter?

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt (hunk 1 above)
Comment (iaik-jheher) [reply datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt:36]:
```kotlin
val defined = when (it) {
        is CryptoSignature.WellDefined -> it
        is CryptoSignature.EC.IndefiniteLength -> it.withCurve()
    }
```
https://pl.kotl.in/MyyqCQvUs

Comment (JesusMcCloud) [reply datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt:36]: to me, this is not a technical issue, but one of how expressive the code is to a human. without the inverse interface, something the grouping (first code block I posted) won't be possible: https://pl.kotl.in/AX88suTB4

**Comment by JesusMcCloud** [comment]
@nodh I addressed all your comments

File: datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKeyParams.kt (hunk 6 above)
Comment (nodh) [reply datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKeyParams.kt:90]: You did not address this issue imho

**Comment by JesusMcCloud** [comment]
> > @nodh I addressed all your comments
> 
> I don't think so, at least the last commit [b97f224](https://github.com/a-sit-plus/kmp-crypto/commit/b97f224afa0536e79d5d2f192c60f6fff15c4808) doesn't address my comments.

TIL: pushing commits help them get to the remote

**Review by nodh — APPROVED** [review]

File: datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKeyParams.kt [hunk 19]
```diff
@@ -166,21 +162,33 @@ sealed class CoseKeyParams {
...
-                    n = n ?: throw IllegalArgumentException("Missing modulus n"),
-                    e = e?.let { bytes -> Int.decodeFromDer(bytes) }
-                        ?: throw IllegalArgumentException("Missing or invalid exponent e")
-                )
-            }.wrap()
+        override fun toCryptoPublicKey(): KmmResult<CryptoPublicKey> = catching {
+            CryptoPublicKey.Rsa(
+                n = n ?: throw IllegalArgumentException("Missing modulus n"),
+                e = e?.let { bytes -> Int.decodeFromDer(bytes) }
+                    ?: throw IllegalArgumentException("Missing or invalid exponent e")
```
Comment (iaik-jheher) [inline datatypes-cose/src/commonMain/kotlin/at/asitplus/crypto/datatypes/cose/CoseKeyParams.kt:169]:
```suggestion
                e = Int.decodeFromDer(e ?:
                    throw IllegalArgumentException("Missing or invalid exponent e"))
```

File: datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt [hunk 20]
```diff
@@ -137,24 +147,27 @@ sealed class CryptoSignature(
...
+                val max = scalarByteLength.toInt() * 8
 
                 require(r.bitLength() <= max) {
-                    "r is ${r.bitLength()} bits long, expected at most ${scalarByteLength.toInt()} bytes (${max} bits)"}
+                    "r is ${r.bitLength()} bits long, expected at most ${scalarByteLength.toInt()} bytes (${max} bits)"
+                }
 
-                require(s.bitLength() <= scalarByteLength.toInt()*8) {
-                    "s is ${s.bitLength()} bits long, expected at most ${scalarByteLength.toInt()} bytes (${max} bits)"}
+                require(s.bitLength() <= scalarByteLength.toInt() * 8) {
```
Comment (iaik-jheher) [inline datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt:166]:
```suggestion
                require(s.bitLength() <= max) {
```

**Review by iaik-jheher — APPROVED** [review]
